scripts/docs_link_checker.py

```diff
@@ -0,0 +1,306 @@
+            if match:
+                result = self._check_relative_link(link, file, match.group("path"))
+            else:
+                match = self._absolute_link_pattern.match(link)
+                if match:
+                    result = self._check_absolute_link(link, file, match.group("path"))
+                else:
+                    match = self._docroot_link_pattern.match(link)
+                    if match:
+                        result = self._check_docroot_link(
+                            link, file, match.group("path")
+                        )
+                    else:
+                        result = LinkReport(link, file, "Invalid link format")
+
+        return result
+
+    def check_file(self, file: str) -> List[LinkReport]:
+        """Looks for all the links in a file and checks them.
+
+        Returns:
+            A list of broken links, or an empty list if no links are broken
+        """
+        with open(file) as f:
+            contents = f.read()
+
+        matches = self._markdown_link_pattern.finditer(contents)
+
+        result: List[LinkReport] = []
+
+        for match in matches:
+            report = self.check_link(match, file)
+            if report is not None:
+                result.append(report)
+
+        return result
+
+
+@click.command()
+@click.option("--path", "-p", default=".", help="Path to markdown files to check")
+@click.option(
+    "--docs-root", "-r", default=None, help="Root to all docs for link checking"
+)
+@click.option(
+    "--site-prefix",
+    "-s",
+    default="",
+    help="Top-most folder in the docs URL for resolving absolute paths",
+)
+@click.option("--skip-external", is_flag=True)
+def scan_docs(path: str, docs_root: str, site_prefix: str, skip_external: bool):
+    # verify that our path is correct
+    if not os.path.isdir(path):
+        print(f"Docs path: {path} is not a directory")
+        exit(1)
+
+    if docs_root is None:
+        docs_root = path
+    elif not os.path.isdir(docs_root):
+        print(f"Docs root path: {docs_root} is not a directory")
+        exit(1)
+
+    # prepare our return value
+    result: List[LinkReport] = list()
+    checker = LinkChecker(path, docs_root, site_prefix, skip_external)
+
+    for file in glob.glob(f"{path}/**/*.md", recursive=True):
+        report = checker.check_file(file)
+        if report:
+            result.extend(report)
+
+    if result:
+        logger.info("%i broken links found", len(result))
+        print("----------------------------------------------")
+        print("------------- Broken Link Report -------------")
+        print("----------------------------------------------")
+        for line in result:
+            print(line)
+
+        exit(1)
+    else:
+        logger.info("No broken links found")
+        print("No broken links found")
+        exit(0)
+
+
+def main():
+    scan_docs()
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Relates to my first comment but how do I use this script? Do I invoke with `python scripts/docs_link_checker.py`? Should we add a shebang to make this executable with `./scripts/docs_link_checker`?

Reply:
done!